### ISSUE TITLE
Non-Selective Query Error for Opportunity Cascade Delete

### ISSUE THREAD

**Issue by jesslopez-sf** [SalesforceFoundation/NPSP#4864]
This occurs when [OPP_CascadeDeleteLookups_TDTM ](https://github.com/SalesforceFoundation/Cumulus/blob/master/src/classes/OPP_CascadeDeleteLookups_TDTM.cls#L64-L71) queries for GAU Allocations where the Opportunity or the Payment's Opportunity is in the set of deleted Opportunities. In orgs with more than 200,000 relevant records or if an index threshold is hit, this error is returned. I believe we might see this one come up more frequently as orgs might be using a default GAU for all of their Opportunities and/or we have orgs with large volumes that could definitely have skewed GAUs.

```public List<Allocation__c> load(Set<Id> opportunityIds) {
return [
          SELECT Opportunity__c
          FROM Allocation__c
          WHERE Opportunity__c IN :opportunityIds 
               OR Payment__r.npe01__Opportunity__c IN :opportunityIds
      ];
}
```

[We changed this query in 3.148](https://github.com/SalesforceFoundation/Cumulus/commit/2582ca12daa0a7ee588d929c5fc742771a4067b4) to also search in the GAU Allocation's Payment's Opportunity; not sure if that exacerbated this issue.

### Knowledge Article
https://help.salesforce.com/articleView?id=000323572&type=1&mode=1

- Likely Reason #1: "1. You may find that the query in question needs to be more selective in the WHERE clause. According to the Salesforce standards & best practices - the where clause needs to subset 10% or less of the data."
- The WHERE clause currently contains Id fields, though it is a composite filter

### StackTrace:
CANNOT_INSERT_UPDATE_ACTIVATE_ENTITY: npsp.TDTM_RecurringDonation: execution of AfterUpdate caused by: System.DmlException: Delete failed. 

First exception on row 0 with id 006xxxxxxxxxxxxxxx; 

first error: CANNOT_INSERT_UPDATE_ACTIVATE_ENTITY, npsp.TDTM_Opportunity: execution of BeforeDelete caused by: System.QueryException: 

Non-selective query against large object type (more than 200000 rows). Consider an indexed filter or contact salesforce.com about custom indexing. Even if a field is indexed a filter might still not be selective when: 

1. The filter value includes null (for instance binding with a list that contains null) 
2. Data skew exists whereby the number of matching rows is very large (for instance, filtering for a particular foreign key value that occurs many times) 

```
Class.npsp.OPP_CascadeDeleteLookups_TDTM.CascadeDeleteLoader.load: line 65, column 1 
External entry point Class.npsp.CDL_CascadeDeleteLookups.CascadeDelete.load: line 175, column 1 
Class.npsp.CDL_CascadeDeleteLookups_TDTM.runBeforeDelete: line 91, column 1 
Class.npsp.CDL_CascadeDeleteLookups_TDTM.run: line 68, column 1 
Class.npsp.TDTM_TriggerHandler.runClass: line 193, column 1 
Class.npsp.TDTM_TriggerHandler.run: line 113, column 1 
Class.npsp.TDTM_Config_API.run: line 92, column 1 
Trigger.npsp.TDTM_Opportunity: line 33, column 1: [] 
```

The context for the customer Case is a Recurring Donation update causing all related Opportunities to be deleted and reinserted (this happens with any update to fixed-length RDs or if an open-ended RD's Installment Period or Contact changes). Specifically in that Case, this is the initial stacktrace that led to the error above:

```
Class.npsp.RD_RecurringDonations.handleRecurringDonationUpdate: line 869, column 1 
Class.npsp.RD_RecurringDonations_TDTM.run: line 109, column 1 
Class.npsp.TDTM_TriggerHandler.runClass: line 193, column 1 
Class.npsp.TDTM_TriggerHandler.run: line 113, column 1 
Class.npsp.TDTM_Config_API.run: line 92, column 1 
Trigger.npsp.TDTM_RecurringDonation: line 33, column 1
```

### Steps To Repeat
Likely the fastest way would be to create an org with 1 GAU and have all >200,000 Opportunities associated with that GAU via GAU Allocations. I ran into this same issue in a test org where I deliberately had skew; the majority of my GAU Allocations were soft deleted and I still ran into this error even when attempting to delete a single Opportunity until I hard deleted the GAU Allocations.

1. Create GAU
2. Create 200,000 Opportunities with GAU Allocations for that GAU
3. Delete 1 or more Opportunities

**Comment by jesslopez-sf** [comment]
**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001HpWTQA0" target="_blank">W-037942</a>

**Comment by ravensisco** [comment]
Included in beta release 1.0 (Beta 2)
